ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,435 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
+
+    DPRINT("Symbolic link without prefix is '%wZ'\n", &LinkNameNoPrefix);
 
-    if (InstanceKey != NULL)
-        InstanceKeyRealP = InstanceKey;
+    /* Find the reference string, if any */
+    for (i = 0; i < LinkNameNoPrefix.Length / sizeof(WCHAR); i++)
+    {
+        if (LinkNameNoPrefix.Buffer[i] == L'\\')
+        {
+            break;
+        }
+    }
+    ReferenceStringOffset = i * sizeof(WCHAR);
+
+    /* The GUID is before the reference string or at the end */
+    ASSERT(LinkNameNoPrefix.Length >= ReferenceStringOffset);
+    if (ReferenceStringOffset < GUID_STRING_BYTES + sizeof(WCHAR))
+    {
+        DPRINT1("Invalid link name '%wZ'\n", SymbolicLinkName);
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    /* Get reference string (starts after {GUID} and \) from link without prefix */
+    // TODO: Add some more sanity checks
+    ReferenceStringReal.Buffer = LinkNameNoPrefix.Buffer + ReferenceStringOffset / sizeof(WCHAR);
+    ReferenceStringReal.Length = LinkNameNoPrefix.Length - ReferenceStringOffset;
+    ReferenceStringReal.MaximumLength = ReferenceStringReal.Length;
+
+    /* Get GUID string (device class GUID in {} brackets) */
+    GuidStringReal.Buffer = LinkNameNoPrefix.Buffer + (ReferenceStringOffset - GUID_STRING_BYTES) / sizeof(WCHAR);
+    GuidStringReal.Length = GUID_STRING_BYTES;
+    GuidStringReal.MaximumLength = GuidStringReal.Length;
+
+    /* Create a temporary device string (looks like ACPI#PNP0501#1#) */
+    DeviceString.Buffer = LinkNameNoPrefix.Buffer;
+    DeviceString.Length = LinkNameNoPrefix.Length - ReferenceStringReal.Length;
+    DeviceString.MaximumLength = DeviceString.Length;
+
+    DPRINT("Device string is '%wZ'\n", &DeviceString);
+
+    if (MungedString != NULL)
+    {
+        /* Allocate a munged path string */
+        MungedStringReal.Length = 0;
+        MungedStringReal.MaximumLength = MungedPrefix.Length + DeviceString.Length + sizeof(UNICODE_NULL);
+        MungedStringReal.Buffer = ExAllocatePoolWithTag(PagedPool, MungedStringReal.MaximumLength, TAG_IO);
+        if (!MungedStringReal.Buffer)
+        {
+            DPRINT1("ExAllocatePoolWithTag() failed\n");
+            return STATUS_INSUFFICIENT_RESOURCES;
+        }
+
+        /* Fill it by munged prefix and our previously created device string */
+        Status = RtlAppendUnicodeStringToString(&MungedStringReal, &MungedPrefix);
+        NT_ASSERT(NT_SUCCESS(Status));
+
+        Status = RtlAppendUnicodeStringToString(&MungedStringReal, &DeviceString);
+        NT_ASSERT(NT_SUCCESS(Status));
+
+        DPRINT("Munged string is '%wZ'\n", &MungedStringReal);
+    }
+
+    DPRINT("GUID string is '%wZ'\n", &GuidStringReal);
+    DPRINT("Reference string is '%wZ'\n", &ReferenceStringReal);
+
+    /* Store received parts if the parameters are not null */
+    if (PrefixString != NULL)
+    {
+        if (RtlPrefixUnicodeString(&KernelModePrefix, SymbolicLinkName, FALSE))
```

Comment:
There should be a way, at the beginning of the function, where you also check whether these prefixes are present in the symbolic name (if not you error out), to also remember which one is present, so that here you could just directly assign PrefixString to the correct values. (Also it should be noted that since both prefixes have the same length, one could simplify the code here...)

Reply:
Yes, this is indeed better. And I did it from the start, although not at the beginning of the function. But it was @ThFabba's suggestion to assign each member of the structure separately. I though there was a reason why he suggested it. Your suggestion, in essence, reverts the Thomas's one. :sweat_smile:

Reply:
Are you referring to this comment? https://github.com/reactos/reactos/pull/3510#discussion_r744136334
This was more about where the buffer is pointing (it should be inside the caller's `SymbolicLinkName` buffer, not one of your constants).
My code snippet there actually does exactly what Hermès is suggesting wrt simplifying because they have the same length.

You could do something like
```
C_ASSERT(KernelModePrefix.Length == UserModePrefix.Length);
ASSERT(RtlPrefixUnicodeString(&KernelModePrefix, SymbolicLinkName, FALSE) ||
       RtlPrefixUnicodeString(&UserModePrefix, SymbolicLinkName, FALSE));
```
(and/or a comment) to make this assumption clear.

Reply:
> Are you referring to this comment? https://github.com/reactos/reactos/pull/3510#discussion_r744136334

Yes, Thomas, I indeed meant that. Thanks for pointing out. :slightly_smiling_face:

Reply:
I don't think the ASSERT(RtlPrefix...) is wanted, but instead a dynamic check and fail, as currently being done (otherwise passing an invalid string from usermode, e.g. fuzzing, would make the code crash) . The C_ASSERT on the contrary should be fine.
Then, my suggestion for that `if (PrefixString != NULL)` code block, would be that the if(RtlPrefix...) + assignments get simplified to a single block
```
PrefixString->Buffer = SymbolicLinkName->Buffer;
PrefixString->Length = KernelModePrefix.Length; // Same as UserModePrefix.Length
PrefixString->MaximumLength = PrefixString->Length;
```

Reply:
So? Should I apply Thomas's suggestion here?